ambry-tools/src/main/java/com.github.ambry/clustermap/HelixBootstrapUpgradeTool.java

```diff
@@ -463,8 +476,179 @@ private void addNewAmbryPartitions(List<Partition> partitions, String resourceNa
     return replicaList;
   }
 
+  /**
+   * Get the instance name string associated with this data node in Helix.
+   * @param dataNode the {@link DataNodeId} of the data node.
+   * @return the instance name string.
+   */
   private static String getInstanceName(DataNodeId dataNode) {
     return dataNode.getHostname() + "_" + dataNode.getPort();
   }
+
+  void validateAndClose() {
+    try {
+      verifyEquivalencyWithStatic(staticClusterMap.hardwareLayout, staticClusterMap.partitionLayout);
+    } finally {
+      for (HelixAdmin admin : adminForDc.values()) {
+        admin.close();
+      }
+    }
+  }
+
+  /**
+   * Verify that the information in Helix and the information in the static clustermap are equivalent.
+   * @param hardwareLayout the {@link HardwareLayout} of the static clustermap.
+   * @param partitionLayout the {@link PartitionLayout} of the static clustermap.
+   */
+  private void verifyEquivalencyWithStatic(HardwareLayout hardwareLayout, PartitionLayout partitionLayout) {
+    String clusterName = hardwareLayout.getClusterName();
+    for (Datacenter dc : hardwareLayout.getDatacenters()) {
+      HelixAdmin admin = adminForDc.get(dc.getName());
+      ensureOrThrow(admin != null, "No ZkInfo for datacenter " + dc.getName());
+      ensureOrThrow(admin.getClusters().contains(clusterName),
+          "Cluster not found in ZK " + dataCenterToZkAddress.get(dc.getName()));
+      verifyResourcesAndPartitionEquivalencyInDc(dc, clusterName, partitionLayout);
+      verifyDataNodeAndDiskEquivalencyInDc(dc, clusterName, partitionLayout);
+    }
+  }
+
+  /**
+   * Verify that the hardware layout information is in sync - which includes the node and disk information. Also verify
+   * that the replicas belonging to disks are in sync between the static cluster map and Helix.
+   * @param dc the datacenter whose information is to be verified.
+   * @param clusterName the cluster to be verified.
+   * @param partitionLayout the {@link PartitionLayout} of the static clustermap.
+   */
+  private void verifyDataNodeAndDiskEquivalencyInDc(Datacenter dc, String clusterName,
```

Comment:
I assume these verify methods are copied from the test class and hence I am not reviewing once again. If there are many changes, let me know